File: backend/app/schemas/analytics.py
"""Analytics ingest schemas.

The ingest endpoint is the one place the app accepts arbitrary client-supplied
structure, so the caps here ARE the security boundary. An events table with no
bounds is both a storage bomb and a place for user-typed text (chat messages,
search queries) to leak into a store the privacy policy never described.

Three defences, all enforced before anything reaches Postgres:
  1. `event` must be on `ALLOWED_EVENTS` — unknown names are DROPPED, not stored.
     Analytics is a fixed vocabulary; anything else is a client bug or an abuse.
  2. Batch size, prop count, key length, and value length are all capped.
  3. Prop VALUES are scalars only. No nested objects/arrays — that's how an entire
     request body ends up in a "dimension" by accident.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, field_validator

# One HTTP round-trip per app background, not per event. A caller sending more than
# this in one batch is a bug or an abuse, not a session's worth of activity.
MAX_EVENTS_PER_BATCH = 50

# Per-event prop bounds. Dimensions are things like ticker/persona/screen — a handful
# of short scalars. Anything larger is text that does not belong here.
MAX_PROPS_PER_EVENT = 12
MAX_PROP_KEY_CHARS = 40
MAX_PROP_VALUE_CHARS = 120
MAX_EVENT_NAME_CHARS = 60
MAX_SESSION_ID_CHARS = 64

# The fixed analytics vocabulary. Adding an event means adding it HERE and on the iOS
# side — deliberately not open-ended, so a typo silently dropping data is caught by
# the parity test rather than becoming a permanently empty column in a dashboard.
#
# Chosen to answer the launch questions that currently have no answer at all:
# activation (install → first report), conversion (paywall → purchase), engagement
# (which tabs/personas), and retention (app_open over time).
ALLOWED_EVENTS: frozenset[str] = frozenset({
    "app_open",
    "screen_view",
    "report_requested",
    "report_completed",
    "report_failed",
    "chat_sent",
    "paywall_shown",
    "paywall_purchase_started",
    "purchase_completed",
    "watchlist_added",
    "lesson_completed",
    "audio_played",
    # First-run funnel. `onboarding_completed` carries the ticker count, which is
    # THE activation number for this app: a populated watchlist is what makes
    # Updates, the personalized Home strip, and (later) push relevant at all.
    "onboarding_completed",
    "onboarding_skipped",
    # Failure telemetry. Both replace `print`/`#if DEBUG` sites that were invisible in release
    # builds, which is how a whole class of silently-reverted user actions (whale follow, the
    # star toggles, portfolio edits) survived unnoticed. `mutation_failed` is user-initiated,
    # `background_sync_failed` is best-effort sync (settings, push registration, guest claim).
    "mutation_failed",
    "background_sync_failed",
})


# Prop KEYS are allowlisted too, not just values. `/events` is public and
# unauthenticated, so without this the DB-level guarantee would be "≤12 arbitrary
# keys × ≤120 chars of caller-supplied text per event" — which is not what the
# privacy policy or the App Privacy filing describes. Every key here is a
# low-cardinality dimension chosen at an iOS call site, never user input.
ALLOWED_PROP_KEYS: frozenset[str] = frozenset({
    "tab",        # screen_view — fixed 5-case HomeTab enum
    "ticker",     # report_*, watchlist_added
    "persona",    # report_* — fixed persona key
    "tier",       # paywall_*, purchase_completed
    "reason",     # report_failed — AppError.analyticsCode, never the message
    "kind",       # lesson_completed, audio_played — fixed category
    "context",    # chat_sent — ChatContextType enum, never the message body
    "seconds",    # durations
    "count",
    "cached",
    "action",     # mutation_failed — a fixed infinitive phrase from the call site
    "op",         # background_sync_failed — fixed operation name
    "code",       # both — AppError.analyticsCode, never the message
})


class AnalyticsEvent(BaseModel):
    """One client-emitted event."""

    event: str
    props: Dict[str, Any] = Field(default_factory=dict)
    # Device clock. Recorded for intra-session ordering, never trusted for bucketing —
    # the server stamps `server_ts` and every aggregate uses that.
    client_ts: Optional[str] = None

    @field_validator("client_ts")
    @classmethod
    def _parseable_timestamp(cls, v: Optional[str]) -> Optional[str]:
        """Null anything Postgres would reject.

        This column is TIMESTAMPTZ, and a single unparseable value fails the ENTIRE
        multi-row INSERT (SQLSTATE 22007) — so one client with a broken clock format
        could silently destroy 49 other good events per batch, repeatedly.
        """
        if not v:
            return None
        try:
            datetime.fromisoformat(v.replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return None
        return v

    @field_validator("event")
    @classmethod
    def _trimmed(cls, v: str) -> str:
        # TRUNCATE, never `max_length`. A length constraint RAISES, which the global
        # RequestValidationError handler turns into a 422 — discarding all 50 events
        # in the batch. That is the exact failure `is_known` exists to avoid, so an
        # over-long name must degrade the same way an unknown one does: drop the one
        # event (it won't match the allowlist), keep the rest.
        return (v or "").strip()[:MAX_EVENT_NAME_CHARS]

    @property
    def is_known(self) -> bool:
        """Whether this event survives the allowlist.

        Deliberately NOT a validator that raises: a rejected field fails the whole
        request, so one unrecognised name would discard every event in the batch
        alongside it. That is a live scenario — an app update ships a new event
        before the backend allowlist learns it — and losing a whole flush of good
        data to one unknown name is far worse than dropping the one.
        """
        return self.event in ALLOWED_EVENTS

    @field_validator("props")
    @classmethod
    def _bounded_scalar_props(cls, v: Dict[str, Any]) -> Dict[str, Any]:
        """Keep only short scalar dimensions. Silently drops the rest rather than
        rejecting the whole event — losing one malformed dimension is much better
        than losing the event (and analytics must never break the app)."""
        if not v:
            return {}
        cleaned: Dict[str, Any] = {}
        # Filter FIRST, cap after. Slicing first meant a payload leading with 12 junk
        # keys silently discarded the real dimension behind them.
        for key, value in v.items():
            if len(cleaned) >= MAX_PROPS_PER_EVENT:
                break
            if not isinstance(key, str) or not key or len(key) > MAX_PROP_KEY_CHARS:
                continue
            # Key ALLOWLIST. The privacy justification for this whole table is that
            # props carry dimensions, never free text — with arbitrary keys accepted
            # that was a convention the server did not enforce, on a public
            # unauthenticated endpoint. Now it does.
            if key not in ALLOWED_PROP_KEYS:
                continue
            if isinstance(value, bool) or isinstance(value, (int, float)):
                cleaned[key] = value
            elif isinstance(value, str):
                cleaned[key] = value[:MAX_PROP_VALUE_CHARS]
            # dict / list / None → dropped: a nested payload is how a whole request
            # body (or a user's typed message) ends up in a "dimension".
        return cleaned


class AnalyticsBatchRequest(BaseModel):
    """Body of ``POST /api/v1/events``."""

    events: List[AnalyticsEvent] = Field(default_factory=list)
    session_id: Optional[str] = None

    @field_validator("session_id")
    @classmethod
    def _bounded_session_id(cls, v: Optional[str]) -> Optional[str]:
        # Truncate, don't raise — same reason as `event` above.
        return v[:MAX_SESSION_ID_CHARS] if v else None

    @field_validator("events")
    @classmethod
    def _bounded_batch(cls, v: List[AnalyticsEvent]) -> List[AnalyticsEvent]:
        return v[:MAX_EVENTS_PER_BATCH]


class AnalyticsBatchResponse(BaseModel):
    """`accepted` can be lower than what was sent — unknown event names and
    over-cap events are dropped, not errors. The client is fire-and-forget and
    ignores this; it exists so the drop rate is observable from curl/tests."""

    accepted: int
    dropped: int = 0

File: frontend/ios/ios/Core/Services/Analytics.swift
//
//  Analytics.swift
//  ios
//
//  First-party product analytics. Buffers events and flushes them to
//  `POST /api/v1/events`.
//
//  WHY FIRST-PARTY (no PostHog / Amplitude / TelemetryDeck)
//  --------------------------------------------------------
//  The App Privacy questionnaire and both privacy-policy surfaces were finalised
//  before launch (documents/legal/app-privacy-answers.md — 8 data types,
//  tracking = No). Adding a third-party analytics SDK reopens all of that and adds
//  a processor to disclose. Events into our own backend disclose nothing new.
//
//  RULES THIS FILE ENFORCES
//  ------------------------
//  1. **Never affects the app.** Every call is fire-and-forget. Failures are
//     swallowed. `track` returns instantly — it only appends to an in-memory buffer.
//  2. **Never sends user-typed text.** Props are literals chosen at the call site:
//     tickers, persona keys, screen names, tiers. Never a chat message, a search
//     query, an email, or an error body. The backend independently enforces an
//     allowlist and size caps, but the first line of defence is here.
//  3. **Batched, not chatty.** Events flush on background, on a full buffer, or on a
//     timer — not one request per event.
//
//  Identity is server-side: the backend derives the bucket from the `X-Guest-Id`
//  header APIClient already sends (or the Bearer token when signed in), so nothing
//  identifying is assembled here.
//

import Foundation
import OSLog
import UIKit

/// The fixed analytics vocabulary. Must stay in sync with `ALLOWED_EVENTS` in
/// `backend/app/schemas/analytics.py` — the backend DROPS unknown names (it does not
/// reject the batch), so a drift here shows up as a silently empty metric.
enum AnalyticsEventName: String {
    case appOpen = "app_open"
    case screenView = "screen_view"
    case reportRequested = "report_requested"
    case reportCompleted = "report_completed"
    case reportFailed = "report_failed"
    case chatSent = "chat_sent"
    case paywallShown = "paywall_shown"
    case paywallPurchaseStarted = "paywall_purchase_started"
    case purchaseCompleted = "purchase_completed"
    /// Consumable credit packs ("Buy Credits"). Separate from the `paywall*` trio on purpose:
    /// a top-up and a subscription are different decisions, and merging their funnels would
    /// make the subscription conversion rate — the number the pricing model rests on —
    /// unreadable. Carry `product_id`, and `count` for the credits granted.
    case creditPackShown = "credit_pack_shown"
    case creditPackPurchaseStarted = "credit_pack_purchase_started"
    case creditPackPurchased = "credit_pack_purchased"
    case watchlistAdded = "watchlist_added"
    case lessonCompleted = "lesson_completed"
    case audioPlayed = "audio_played"
    /// First-run funnel. `onboardingCompleted` carries `count` — the activation
    /// metric, since a populated watchlist is what makes Updates, the Home strip,
    /// and push relevant at all.
    case onboardingCompleted = "onboarding_completed"
    case onboardingSkipped = "onboarding_skipped"
    /// Personalized explanations opted in / out. Deliberately property-less: the point
    /// is the RATE and the consent audit trail, and the answers themselves are the
    /// reader's own reading interests, which have no business in the analytics stream.
    case personalizationConsentGranted = "personalization_consent_granted"
    case personalizationConsentWithdrawn = "personalization_consent_withdrawn"
    /// A user-initiated write that failed and was reverted. Carries `action` (what they were
    /// doing) and `code` (`AppError.analyticsCode`). Exists because this whole class of failure
    /// used to be a `print` — so in production there was no way to know how often a tap on
    /// Follow, a star, or a portfolio edit quietly did nothing.
    case mutationFailed = "mutation_failed"

    /// Push funnel. Nothing in the app measured notifications at all, so "did anyone tap
    /// it?" and "how many people turned them off?" — the two questions that decide whether
    /// a category is worth sending — were unanswerable.
    ///
    /// `kind` is the fixed registry key ("earnings_upcoming"), `route` the destination
    /// family. NEVER a ticker: `props` is for low-cardinality dimensions, and a per-user
    /// value there is both useless as a metric and a privacy footgun.
    case pushReceived = "push_received"
    case pushOpened = "push_opened"
    case pushPermissionResult = "push_permission_result"
    case pushRegisterFailed = "push_register_failed"
    case notificationInboxOpened = "notification_inbox_opened"
    case priceAlertCreated = "price_alert_created"
    /// A best-effort background sync that failed (settings hydrate/push, device registration,
    /// guest-data claim). Same rationale: these were `#if DEBUG` prints, i.e. invisible in the
    /// builds that matter.
    case backgroundSyncFailed = "background_sync_failed"
}

/// A single buffered event. `props` is deliberately `[String: AnalyticsValue]` rather
/// than `[String: Any]` so a non-scalar can't be passed by accident.
struct BufferedEvent: Encodable, Sendable {
    let event: String
    let props: [String: AnalyticsValue]
    let clientTs: String

    enum CodingKeys: String, CodingKey {
        case event, props
        case clientTs = "client_ts"
    }
}

/// Scalar-only prop values. The type system is the enforcement: there is no case for
/// a dictionary or array, so a nested payload cannot be attached to an event.
enum AnalyticsValue: Encodable, Sendable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let v): try c.encode(v)
        case .int(let v):    try c.encode(v)
        case .double(let v): try c.encode(v)
        case .bool(let v):   try c.encode(v)
        }
    }
}

extension AnalyticsValue: ExpressibleByStringLiteral,
                          ExpressibleByIntegerLiteral,
                          ExpressibleByFloatLiteral,
                          ExpressibleByBooleanLiteral {
    init(stringLiteral value: String)  { self = .string(value) }
    init(integerLiteral value: Int)    { self = .int(value) }
    init(floatLiteral value: Double)   { self = .double(value) }
    init(booleanLiteral value: Bool)   { self = .bool(value) }
}

/// `nonisolated`: encoded by `APIClient` (an actor) off the main actor, so its `Encodable`
/// conformance must not be main-actor-isolated. Already `Sendable`; this makes it honest.
nonisolated struct AnalyticsBatchRequest: Encodable, Sendable {
    let events: [BufferedEvent]
    let sessionId: String

    enum CodingKeys: String, CodingKey {
        case events
        case sessionId = "session_id"
    }
}

actor Analytics {
    static let shared = Analytics()

    /// Matches `MAX_EVENTS_PER_BATCH` on the backend. Flushing at the cap keeps a busy
    /// session from ever exceeding what the server will accept in one request.
    private static let batchSize = 50
    /// Hard ceiling on the buffer so a long offline stretch can't grow memory without
    /// bound. Oldest events are dropped first — recent behaviour is the useful part.
    private static let maxBuffered = 500
    private static let flushInterval: TimeInterval = 30

    /// Per-launch, not per-install: lets events be grouped into a session without
    /// carrying any identity of its own (identity is server-derived).
    private let sessionId = UUID().uuidString

    private static let log = Logger(subsystem: "com.phan.caydex", category: "analytics")

    private var buffer: [BufferedEvent] = []
    private var flushTask: Task<Void, Never>?
    private var isFlushing = false

    private let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    // MARK: - Public API

    /// Record an event. Returns immediately; nothing here can throw or block.
    nonisolated func track(
        _ event: AnalyticsEventName,
        _ props: [String: AnalyticsValue] = [:]
    ) {
        Task { await self.enqueue(event, props) }
    }

    /// Flush now. Called on `scenePhase == .background` — the moment the app is most
    /// likely to be suspended with buffered events still pending.
    ///
    /// Held open by a background-task assertion, because "about to be suspended" is exactly
    /// when an un-asserted request loses its race: the buffer is an in-memory actor property
    /// with NO disk persistence, so a process frozen mid-flush loses the batch outright, with
    /// nothing to resume from on next launch.
    nonisolated func flushNow() {
        Task { @MainActor in
            let app = UIApplication.shared
            var assertion: UIBackgroundTaskIdentifier = .invalid
            assertion = app.beginBackgroundTask(withName: "caydex.analytics.flush") {
                guard assertion != .invalid else { return }
                app.endBackgroundTask(assertion)
                assertion = .invalid
            }
            await self.flush()
            guard assertion != .invalid else { return }
            app.endBackgroundTask(assertion)
            assertion = .invalid
        }
    }

    // MARK: - Internals

    private func enqueue(_ event: AnalyticsEventName, _ props: [String: AnalyticsValue]) {
        buffer.append(BufferedEvent(
            event: event.rawValue,
            props: props,
            clientTs: iso.string(from: Date())
        ))

        if buffer.count > Self.maxBuffered {
            // Drop OLDEST. A long offline stretch should cost the stale tail, not the
            // events that just happened.
            buffer.removeFirst(buffer.count - Self.maxBuffered)
        }

        if buffer.count >= Self.batchSize {
            Task { await self.flush() }
        } else {
            scheduleFlush()
        }
    }

    private func scheduleFlush() {
        guard flushTask == nil else { return }
        flushTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(Self.flushInterval * 1_000_000_000))
            guard !Task.isCancelled else { return }
            // MUST be `flushFromTimer`, never `flush()` directly — see below.
            await self?.flushFromTimer()
        }
    }

    /// The timer's OWN entry point into `flush()`.
    ///
    /// `flush()` opens by cancelling `flushTask` to kill a pending timer. That is right for the
    /// other two callers — the batch-cap path and `flushNow()` each run in a FRESH task, so the
    /// only thing cancelled is the sleeping timer. From inside the timer task it was suicide:
    /// `flushTask` still referenced the task that was executing, so `flush()` cancelled itself,
    /// and `URLSession.data(for:)` — which is cancellation-aware — failed instantly with
    /// `URLError.cancelled` (-999) before the request ever left the device.
    ///
    /// Every 30-second flush therefore failed, and the batch was dropped by the `catch` below.
    /// Since a normal session buffers well under `batchSize` events and is killed rather than
    /// backgrounded cleanly, that was very nearly ALL analytics — while looking healthy, because
    /// the drop was reported only in DEBUG.
    ///
    /// Releasing the slot instead of cancelling it is the whole fix.
    private func flushFromTimer() async {
        flushTask = nil
        await flush()
    }

    private func flush() async {
        flushTask?.cancel()
        flushTask = nil

        // One flush at a time: two concurrent flushes would either double-send the
        // same events or interleave and lose some on failure.
        guard !isFlushing, !buffer.isEmpty else { return }
        isFlushing = true
        defer { isFlushing = false }

        let batch = Array(buffer.prefix(Self.batchSize))
        buffer.removeFirst(batch.count)

        do {
            let _: AnalyticsBatchResponse = try await APIClient.shared.request(
                endpoint: .trackEvents(
                    AnalyticsBatchRequest(events: batch, sessionId: sessionId)
                ),
                responseType: AnalyticsBatchResponse.self
            )
        } catch {
            // Cancellation is NOT a delivery failure — the request never reached the network,
            // so the retry-amplifier argument below does not apply to it. Put the batch back.
            // Re-inserting at the FRONT restores the original order, which means the bound
            // must now trim the TAIL (the stale end), the opposite end from `enqueue`.
            if Task.isCancelled || (error as? URLError)?.code == .cancelled
                || (AppError.from(error)).isCancellation {
                buffer.insert(contentsOf: batch, at: 0)
                // Same overflow policy as `enqueue`: drop the OLDEST. Re-inserting at the
                // front restores chronological order, so the oldest are once again at the
                // front — which means an overflow here discards part of the batch we just put
                // back. That is correct and deliberate: 500 buffered events means a long
                // offline stretch, and recent behaviour is the useful part.
                if buffer.count > Self.maxBuffered {
                    buffer.removeFirst(buffer.count - Self.maxBuffered)
                }
                Self.log.warning("analytics flush cancelled — \(batch.count, privacy: .public) event(s) re-queued")
                return
            }
            // Deliberately NOT re-queued. Retrying telemetry during a backend outage
            // turns every client into a retry amplifier against an already-struggling
            // server, and stale events aren't worth that. Analytics is lossy by design.
            //
            // Logged at `warning` in EVERY configuration, not just DEBUG. The self-cancelling
            // timer above went unnoticed precisely because this was a `#if DEBUG print`:
            // in production, analytics failing 100% of the time looked exactly like analytics
            // working. Silent degradation is the hardest bug to find.
            Self.log.warning(
                "analytics flush of \(batch.count, privacy: .public) event(s) dropped — \(String(describing: type(of: error)), privacy: .public)"
            )
        }
    }
}

/// `nonisolated`: decoded inside the `Analytics` actor — see `AnalyticsBatchRequest`.
nonisolated struct AnalyticsBatchResponse: Decodable, Sendable {
    let accepted: Int
    let dropped: Int
}
